keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1080 @@
+ }
+ 
+ // NodeConfigStatus describes the status of the config assigned by Node.Spec.ConfigSource.
+```
+
+### CRI API
+
+Add `bool recursive_read_only` to the [`Mount`](https://github.com/kubernetes/cri-api/blob/v0.26.1/pkg/apis/runtime/v1/api.proto#L212-L224) message.
+CRI implementations will also expose the availability of the feature via the `RuntimeHandlerFeatures` message.
+
+As kubelet can inspect the availability of the feature via the `RuntimeHandlerFeatures` message,
+there is no concept of "IfPossible" in the CRI API;
+kubelet translates an "IfPossible" value in the Core API into true or false in the CRI API
+
+The `RuntimeHandlerFeatures` message is also propagated to the `NodeSystemInfo` struct of the Core API.
+
+Diff:
+```diff
+diff --git a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+index 11d58636369..3bbeeca4cb6 100644
+--- a/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
++++ b/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto
+@@ -217,6 +217,8 @@ message Mount {
+     // follow the symlink and mount the real destination to container.
+     string host_path = 2;
+     // If set, the mount is read-only.
++    // In the CRI API, readonly always defaults to false, regardless to the value of
++    // recursive_read_only.
+     bool readonly = 3;
+     // If set, the mount needs SELinux relabeling.
+     bool selinux_relabel = 4;
+@@ -226,6 +228,15 @@ message Mount {
+     repeated IDMapping uidMappings = 6;
+     // GidMappings specifies the runtime GID mappings for the mount.
+     repeated IDMapping gidMappings = 7;
++    // If set to true, the mount is made recursive read-only.
++    // In this CRI API, recursive_read_only is a plain true/false boolean, although its equivalent
++    // in the Kubernetes core API is a quaternary that can be nil, "Enabled", "IfPossible", or "Disabled".
++    // kubelet translates that quaternary value in the core API into a boolean in this CRI API.
++    // Remarks:
++    // - nil is just treated as false
++    // - when set to true, readonly must be explicitly set to true, and propagation must be PRIVATE (0).
++    // - (readonly == false && recursive_read_only == false) does not make the mount read-only.
++    bool recursive_read_only = 8;
+ }
+ 
+ // IDMapping describes host to container ID mappings for a pod sandbox.
+@@ -1512,6 +1523,15 @@ message StatusRequest {
+     bool verbose = 1;
+ }
+ 
++// RuntimeHandlerFeatures describes the features that are supported by the runtime handler.
++message RuntimeHandlerFeatures {
++    // recursive_read_only_mounts is set to true if the runtime handler supports
++    // recursive read-only mounts.
++    // For runc-compatible runtimes, availability of this feature can be detected by checking whether
++    // the Linux kernel version is >= 5.12, and,  `runc features | jq .mountOptions` contains "rro".
++    bool recursive_read_only_mounts = 1;
```

Comment:
with this new pattern we will keep accumulating flags that needs to be returned with every new version of a runtime. When we will get to the point when this feature exists in every runtime with every supported distro, we need to have a mechanism to get rid of this field.

Reply:
the only waty to get rid of fields is to version and advance APIs.  Kubernetes has largely promised not to EOL APIs, so while I truly understand your concern, we can't really do that.

Reply:
I was thinking along the lines of defining this as text-based contract returning set of features "beyond" the default set. And have a definition of a default set as part of CRI version. This way in a few years we can tell that the default set has advanced to include the `recursive_read_only_mounts` so runtimes should not report that they support it any longer.

It changes the shape, performance, and ease of use of this API in short term though. And doesn't make it natural to have a truly-optional features. But solves the problem of deprecation.

CC: @mrunalp @saschagrunert

Reply:
Would it help to make `bool recursive_read_only = 8;` above a union which supports all three types of availability? This way we could get rid of `RuntimeHandlerFeatures` at all.

Reply:
> Would it help to make `bool recursive_read_only = 8;` above a union which supports all three types of availability? This way we could get rid of `RuntimeHandlerFeatures` at all.

Yes, but then we can't show the feature availability in `k get nodes`.

Reply:
If we text based, we will need to define constants somewhere that are shared between kubelet and the runtimes.

Reply:
> I was thinking along the lines of defining this as text-based contract returning set of features "beyond" the default set. 

If this was defined as `repeated string`, it is hard to distinguish "false" from "unknown".

>  When we will get to the point when this feature exists in every runtime with every supported distro, we need to have a mechanism to get rid of this field.

I guess this mechanism will not be needed for a couple of decades, as the wasted fields will not take huge amount of RAM, but kubelet in 2050-ish may choose to just ignore old fields.